pkg/intel/me/me.go

```diff
@@ -0,0 +1,280 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package me
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"strings"
+
+	"github.com/linuxboot/fiano/pkg/uefi"
+)
+
+// LegacyFlashPartitionTableHeader describes the old flash partition table header
+// in Intel ME binaries.
+type LegacyFlashPartitionTableHeader struct {
+	Padding        [16]uint8 // 16 zeros
+	Marker         uint32    // Always $FPT
+	NumFptEntries  uint32
+	HeaderVersion  uint8
+	EntryVersion   uint8
+	HeaderLength   uint8 // Usually 0x30
+	HeaderChecksum uint8
+	TicksToAdd     uint16
+	TokensToAdd    uint16
+	UMASize        uint32
+	Flags          uint32
+}
+
+func (h LegacyFlashPartitionTableHeader) String() string {
+	var b strings.Builder
+	b.WriteString("Flash partition table:\n")
+	fmt.Fprintf(&b, " Entries       : %d\n", h.NumFptEntries)
+	fmt.Fprintf(&b, " HeaderVersion : 0x%x\n", h.HeaderVersion)
+	fmt.Fprintf(&b, " EntryVersion  : 0x%x\n", h.EntryVersion)
+	fmt.Fprintf(&b, " HeaderLength  : 0x%x\n", h.HeaderLength)
+	fmt.Fprintf(&b, " HeaderChecksum: 0x%x\n", h.HeaderChecksum)
+	fmt.Fprintf(&b, " TicksToAdd    : 0x%x\n", h.TicksToAdd)
+	fmt.Fprintf(&b, " TokensToAdd   : 0x%x\n", h.TokensToAdd)
+	fmt.Fprintf(&b, " UMASize       : 0x%x\n", h.UMASize)
+	fmt.Fprintf(&b, " Flags         : 0x%x\n", h.Flags)
+
+	return b.String()
+}
+
+// FlashPartitionTableHeader describes the new flash partition table header
+// in Intel ME binaries.
+type FlashPartitionTableHeader struct {
+	Marker             uint32 // Always $FPT
+	NumFptEntries      uint32
+	HeaderVersion      uint8 // Only support 2.0
+	EntryVersion       uint8
+	HeaderLength       uint8 // Usually 0x20
+	HeaderChecksum     uint8
+	TicksToAdd         uint16
+	TokensToAdd        uint16
+	UMASizeOrReserved  uint32
+	FlashLayoutOrFlags uint32
+	// Not Present in ME version 7
+	FitcMajor  uint16
+	FitcMinor  uint16
+	FitcHotfix uint16
+	FitcBuild  uint16
+}
+
+func (h FlashPartitionTableHeader) String() string {
+	var b strings.Builder
+
+	b.WriteString("Flash partition table:\n")
+	fmt.Fprintf(&b, " Entries            : %d\n", h.NumFptEntries)
+	fmt.Fprintf(&b, " HeaderVersion      : 0x%x\n", h.HeaderVersion)
+	fmt.Fprintf(&b, " EntryVersion       : 0x%x\n", h.EntryVersion)
+	fmt.Fprintf(&b, " HeaderLength       : 0x%x\n", h.HeaderLength)
+	fmt.Fprintf(&b, " HeaderChecksum     : 0x%x\n", h.HeaderChecksum)
+	fmt.Fprintf(&b, " TicksToAdd         : 0x%x\n", h.TicksToAdd)
+	fmt.Fprintf(&b, " TokensToAdd        : 0x%x\n", h.TokensToAdd)
+	fmt.Fprintf(&b, " UMASizeOrReserved  : 0x%x\n", h.UMASizeOrReserved)
+	fmt.Fprintf(&b, " FlashLayoutOrFlags : 0x%x\n", h.FlashLayoutOrFlags)
+	fmt.Fprintf(&b, " Fitc Version       : %d.%d.%d.%d\n", h.FitcMajor, h.FitcMinor, h.FitcHotfix, h.FitcBuild)
+
+	return b.String()
+}
+
+// FlashPartitionTableEntry describes information of a flash partition table entry.
+type FlashPartitionTableEntry struct {
+	Name           [4]uint8
```

Comment:
You can create a new type to make this easier to deal with:

```
// String wraps around byte array to give us more control over how strings are
// serialized.
type String struct {
	Value [4]uint8
}

func (s *String) String() string {
	return strings.TrimRight(string(s.Value[:]), "\x00")
}
```

For example, see https://github.com/linuxboot/fiano/blob/master/pkg/fmap/fmap.go#L33